File: graph/js/graph_recursive.js


const assert = require("assert");
const fs = require("fs");
const readline = require("readline");
const rlimit = require("./build/Release/rlimit_set_stack");

class Vertex {
	constructor(id, colour) {
		this.id = id;
		this.colour = colour;
		this.next = null;
		this.prev = null;
		this.neighbours = new Set();
	}

	get degree() {
		return this.neighbours.size;
	}

	sudoku() {
		this.prev.next = this.next;
		if (this.next != null) {
			this.next.prev = this.prev;
		}
		for (let v of this.neighbours) {
			let b = v.neighbours.delete(this);
			assert(b);
		}
	}

	induce() {
		let induced = new Graph();
		let m = {};
		Graph.explore(this, induced, m, this.neighbours);
		return induced;
	}
}

class Graph {
	constructor() {
		this.dummy = new Vertex(null);
		this.size = 0;
	}

	get head() {
		return this.dummy.next;
	}

	set head(v) {
		this.dummy.next = v;
	}

	shift(v) {
		let h = this.head;
		if (h != null) {
			v.next = h;
			v.prev = this.dummy;
			h.prev = v;
		}
		this.head = v;
		this.size++;
	}

	duplicate() {
		let dup = new Graph();
		let m = {};
		for (let v = this.head; v != null; v = v.next) {
			Graph.explore(v, dup, m, null);
		}
		return dup;
	}

	static explore(start, g, m, valid) {
		let new_vertex = m[start.id];
		if (new_vertex != undefined) {
			return new_vertex;
		}
		new_vertex = new Vertex(start.id, start.colour);
		start.colour.value = 0;
		m[start.id] = new_vertex;
		for (let v of start.neighbours) {
			if (valid != null && !valid.has(v)) {
				continue;
			}
			let neighbour = Graph.explore(v, g, m, valid);
			new_vertex.neighbours.add(neighbour);
		}
		g.shift(new_vertex);
		return new_vertex;
	}

	static explore_iterative(start, g, m, valid) {
		let stack = [];
		stack.push([ start, null ]);
		while (stack.length > 0) {
			let tuple = stack.pop()
			start = tuple[0]
			let sibling = tuple[1]
			let new_vertex = m[start.id];
			if (new_vertex != undefined) {
				if (sibling != null) {
					sibling.neighbours.add(new_vertex);
				}
				continue;
			}
			new_vertex = new Vertex(start.id, start.colour);
			start.colour.value = 0;
			m[start.id] = new_vertex;
			for (let v of start.neighbours) {
				if (valid != null && !valid.has(v)) {
					continue;
				}
				stack.push([ v, new_vertex ])
			}
			g.shift(new_vertex);
			if (sibling != null) {
				sibling.neighbours.add(new_vertex);
			}
		}
	}

	social_credit(bad) {
		bad.sudoku();
		this.size--;
		for (let v of bad.neighbours) {
			v.sudoku();
			this.size--;
		}
	}

	verify_colouring() {
		for (let v = this.head; v != null; v = v.next) {
			if (v.colour.value == 0) {
				return false;
			}
			for (let u of v.neighbours) {
				if (v.colour.value == u.colour.value) {
					return false;
				}
			}
		}
		return true;
	}

	find_max_degree_vertex() {
		let max = 0;
		let ret = null;
		for (let v = this.head; v != null; v = v.next) {
			let d = v.degree;
			if (d > max) {
				ret = v;
				max = d;
			}
		}
		return ret;
	}

	static magic_f(k, n) {
		return Math.ceil(Math.pow(n, 1 - (1 / (k - 1))));
	}

	static k_ge_log_n(k, n) {
		let x = 1;
		for (let i = 0; i < k; i++) {
			x *= 2;
		}
		return x >= n;
	}

	colour_2(i, j) {
		for (let v = this.head; v != null; v = v.next) {
			if (v.colour.value != 0) {
				continue;
			}
			if (!Graph.colour_2_helper(v, i, j)) {
				return false;
			}
		}
		return true;
	}

	static colour_2_helper(v, i, j) {
		if (v.colour.value == j) {
			return false;
		}
		if (v.colour.value == 0) {
			v.colour.value = i;
			for (let u of v.neighbours) {
				if (!Graph.colour_2_helper(u, j, i)) {
					return false;
				}
			}
		}
		return true;
	}

	static colour_2_helper_iterative(v, i, j) {
		let stack = [];
		stack.push([ v, i, j ]);
		while (stack.length > 0) {
			let tuple = stack.pop();
			v = tuple[0];
			i = tuple[1];
			j = tuple[2];
			if (v.colour.value == j) {
				return false;
			}
			if (v.colour.value == 0) {
				v.colour.value = i;
				for (let u of v.neighbours) {
					stack.push([ u, j, i ]);
				}
			}
		}
		return true;
	}

	colour_b(k, i) {
		if (k == 2) {
			if (this.colour_2(i, i + 1)) {
				//console.log("B2: k = " + k + " ret = 2");
				return 2;
			}
			//console.log("B2: k = " + k + " ret = 0");
			return 0;
		}
		let n = this.size;
		if (Graph.k_ge_log_n(k, n)) {
			let j = 0;
			for (let v = this.head; v != null; v = v.next) {
				v.colour.value = i + j;
				j++;
			}
			//console.log("Bn: k = " + k + " ret = " + n);
			return n;
		}
		while (true) {
			let v = this.find_max_degree_vertex();
			if (v.degree < Graph.magic_f(k, n)) {
				//console.log("- breaking k = " + k + " degree = " + v.degree + " magic = " + Graph.magic_f(k, n));
				break;
			}
			let h = v.induce();
			let j = h.colour_b(k - 1, i);
			if (j == 0) {
				//console.log("Bfail: k = " + k + " ret = 0");
				return 0;
			}
			i += j;
			v.colour.value = i;
			this.social_credit(v);
		}
		let max_degree = 0;
		let max_colour = 0;
		for (let v = this.head; v != null; v = v.next) {
			let seen = new Set();
			for (let e of v.neighbours) {
				seen.add(e.colour.value);
			}
			if (max_degree < v.degree) {
				max_degree = v.degree;
			}
			for (let j = i; true; j++) {
				if (!seen.has(j)) {
					v.colour.value = j;
					if (max_colour < j) {
						max_colour = j;
					}
					break;
				}
			}
		}
		assert(max_colour < max_degree + i + 1);

		let ret = max_colour - i + 1;
		let bound = 2 * k * Graph.magic_f(k, n);
		//console.log("Bend: k = " + k + " ret = " + ret);
		return ret > bound ? 0 : ret;
	}

	colour_c() {
		let i = 1;
		while (this.duplicate().colour_b(1 << i, 1) == 0) {
			i++;
		}
		//console.log("C: i = " + i);

		let l = (1 << (i - 1)) + 1;
		let r = 1 << i;
		while (l < r) {
			let m = Math.floor((l + r) / 2);
			if (this.duplicate().colour_b(m, 1) == 0) {
				l = m + 1;
			} else {
				r = m;
			}
		}
		let k = this.duplicate().colour_b(l, 1);
		assert(k != 0);
		return k;
	}

	static from_file(path, callback) {
		let g = new Graph();
		let rl = readline.createInterface({
			input: fs.createReadStream(path, { encoding: 'utf-8' }),
			crlfDelay: Infinity,
		});
		let m = {};
		rl.on("line", function(line) {
			if (line.startsWith("#")) {
				return;
			}
			let parts = line.split(/\s+/);
			let x = parseInt(parts[0], 10);
			let y = parseInt(parts[1], 10);
			let vx = m[x];
			let vy = m[y];
			if (vx == undefined) {
				vx = new Vertex(x, { value: 0 });
				m[x] = vx;
				g.shift(vx);
			}
			if (vy == undefined) {
				vy = new Vertex(y, { value: 0 });
				m[y] = vy;
				g.shift(vy);
			}
			vx.neighbours.add(vy);
			vy.neighbours.add(vx);
		});
		rl.on("close", function() {
			callback(g);
		});
	}
}

function main(args) {
	rlimit.rlimit_set_stack();
	Graph.from_file(args[0], function(g) {
		let k = g.colour_c();
		console.log("k = " + k);
		let b = g.verify_colouring();
		assert(b);
	});
}

if (!module.parent) {
	main(process.argv.slice(2));
}

File: graph/python/graph_recursive.py
#!/usr/bin/env python3

import sys
import math
import resource
import time
from typing import Set, Dict, List, Generator, Tuple, Callable

iterations: int = 0

class Colour:
	def __init__(self: "Colour") -> None:
		self.value: int = 0

class Vertex:
	def __init__(self: "Vertex", colour: Colour = None) -> None:
		self.next: Vertex = None
		self.prev: Vertex = None
		self.colour: Colour = Colour() if colour is None else colour
		self.neighbours: Set[Vertex] = set()

	def degree(self: "Vertex") -> int:
		return len(self.neighbours)

	def sudoku(self: "Vertex") -> None:
		self.prev.next = self.next
		if self.next is not None:
			self.next.prev = self.prev
		v: Vertex
		for v in self.neighbours:
			v.neighbours.remove(self)

	def induce(self: "Vertex") -> "Graph":
		induced: Graph = Graph()
		m: Dict[Vertex, Vertex] = {}
		#Graph.explore(self, induced, m, self.neighbours, time.time())
		Graph.explore(self, induced, m, self.neighbours, None)
		return induced

class Graph:
	EXPLORE: Callable[[Vertex, "Graph", Dict[Vertex, Vertex], Set[Vertex], float], Vertex] = None
	COLOUR_2_HELPER: Callable[[Vertex, int, int], bool] = None

	"""
	@staticmethod
	def initialize():
		Graph.EXPLORE = Graph.explore
		#Graph.EXPLORE = Graph.explore_iterative
		Graph.COLOUR_2_HELPER = Graph.colour_2_helper
		#Graph.COLOUR_2_HELPER = Graph.colour_2_helper_iterative
	"""

	def __init__(self: "Graph") -> None:
		self.dummy: Vertex = Vertex()
		self.dummy.colour = None
		self.size: int = 0
		self.loop_time: float = 0
		self.loop_count: int = 0
		self.stack_time: float = 0
		self.stack_count: int = 0
		self.call_time: float = 0
		self.call_count: int = 0

	def head(self: "Graph") -> Vertex:
		return self.dummy.next

	def vertices(self: "Graph") -> Generator[Vertex, None, None]:
		v: Vertex = self.head()
		while v is not None:
			yield v
			v = v.next

	def shift(self: "Graph", v: Vertex) -> None:
		h: Vertex = self.head()
		if h is not None:
			v.next = h
			v.prev = self.dummy
			h.prev = v
		self.dummy.next = v
		self.size += 1

	def duplicate(self: "Graph") -> "Graph":
		dup: Graph = Graph()
		m: Dict[Vertex, Vertex] = {}
		v: Vertex
		for v in self.vertices():
			#Graph.explore(v, dup, m, None, time.time())
			Graph.explore(v, dup, m, None, None)
		"""
		print("[debug] loop: {:.3f} / {} = {:.3f}, stack: {:.3f} / {} = {:.3f}, call: {:.3f} / {} = {:.3f}".format(
			dup.loop_time,
			dup.loop_count,
			dup.loop_time / (1 if dup.loop_count == 0 else dup.loop_count),
			dup.stack_time,
			dup.stack_count,
			dup.stack_time / (1 if dup.stack_count == 0 else dup.stack_count),
			dup.call_time,
			dup.call_count,
			dup.call_time / (1 if dup.call_count == 0 else dup.call_count),
		))
		"""
		return dup

	@staticmethod
	def explore(start: Vertex, g: "Graph", m: Dict[Vertex, Vertex], valid: Set[Vertex], t_out: float) -> Vertex:
		#t_in: float = time.time()
		#g.call_time += t_in - t_out
		#g.call_count += 1
		new_vertex: Vertex = m.get(start)
		if new_vertex is not None:
			return new_vertex
		new_vertex = Vertex(start.colour)
		start.colour.value = 0
		m[start] = new_vertex
		v: Vertex
		for v in start.neighbours:
			if (valid is not None) and (v not in valid):
				continue
			#neighbour: Vertex = Graph.explore(v, g, m, valid, time.time())
			neighbour: Vertex = Graph.explore(v, g, m, valid, None)
			new_vertex.neighbours.add(neighbour)
		g.shift(new_vertex)
		return new_vertex

	@staticmethod
	def explore_iterative(start: Vertex, g: "Graph", m: Dict[Vertex, Vertex], valid: Set[Vertex], t_out: float) -> None:
		global iterations
		stack: List[Tuple[Vertex, Vertex]] = []
		stack.append((start, None))
		t0: float = time.time()
		while len(stack) > 0:
			iterations += 1
			t1: float = time.time()
			g.loop_time += t1 - t0
			g.loop_count += 1
			sibling: Vertex
			t0 = time.time()
			start, sibling = stack.pop()
			t1 = time.time()
			g.stack_time += t1 - t0
			g.stack_count += 1
			new_vertex: Vertex = m.get(start)
			if new_vertex is not None:
				if sibling is not None:
					sibling.neighbours.add(new_vertex)
				t0 = time.time()
				continue
			new_vertex = Vertex(start.colour)
			start.colour.value = 0
			m[start] = new_vertex
			v: Vertex
			for v in start.neighbours:
				if (valid is not None) and (v not in valid):
					continue
				t0 = time.time()
				stack.append((v, new_vertex))
				t1 = time.time()
				g.stack_time += t1 - t0
				g.stack_count += 1
			g.shift(new_vertex)
			if sibling is not None:
				sibling.neighbours.add(new_vertex)
			t0 = time.time()

	def social_credit(self: "Graph", bad: Vertex) -> None:
		bad.sudoku()
		self.size -= 1
		v: Vertex
		for v in bad.neighbours:
			v.sudoku()
			self.size -= 1

	def verify_colouring(self: "Graph") -> bool:
		v: Vertex
		for v in self.vertices():
			if v.colour.value == 0:
				return False
			u: Vertex
			for u in v.neighbours:
				if v.colour.value == u.colour.value:
					return False
		return True

	def find_max_degree_vertex(self: "Graph") -> Vertex:
		d: int = 0
		ret: Vertex = None
		v: Vertex
		for v in self.vertices():
			e = v.degree()
			if e > d:
				ret = v
				d = e
		return ret

	@staticmethod
	def magic_f(k: int, n: int) -> float:
		return math.ceil(math.pow(n, 1 - (1.0 / (k - 1))))

	@staticmethod
	def k_ge_log_n(k: int, n: int) -> bool:
		return 2 ** k >= n

	def colour_2(self: "Graph", i: int, j: int) -> bool:
		v: Vertex
		for v in self.vertices():
			if v.colour.value != 0:
				continue
			if not Graph.colour_2_helper(v, i, j):
				return False
		return True

	@staticmethod
	def colour_2_helper(v: Vertex, i: int, j: int) -> bool:
		if v.colour.value == j:
			return False
		if v.colour.value == 0:
			v.colour.value = i
			u: Vertex
			for u in v.neighbours:
				if not Graph.colour_2_helper(u, j, i):
					return False
		return True

	@staticmethod
	def colour_2_helper_iterative(v: Vertex, i: int, j: int) -> bool:
		stack: List[Tuple[Vertex, int, int]] = []
		stack.append((v, i, j))
		while len(stack) > 0:
			v, i, j = stack.pop()
			if v.colour.value == j:
				return False
			if v.colour.value == 0:
				v.colour.value = i
				u: Vertex
				for u in v.neighbours:
					stack.append((u, j, i))
		return True

	def colour_b(self: "Graph", k: int, i: int) -> int:
		if k == 2:
			if self.colour_2(i, i + 1):
				#print("B: k = " + str(k) + " ret = 2")
				return 2
			#print("B: k = " + str(k) + " ret = 0")
			return 0
		n: int = self.size
		if Graph.k_ge_log_n(k, n):
			j: int = 0
			v: Vertex
			for v in self.vertices():
				v.colour.value = i + j
				j += 1
			#print('B: k = ' + str(k) + ' ret = ' + str(n))
			return n
		while True:
			v = self.find_max_degree_vertex()
			if v.degree() < Graph.magic_f(k, n):
				#print("- breaking k = {} degree = {} magic = {}".format(k, v.degree(), Graph.magic_f(k, n)));
				break
			"""
			if v.degree() < (k * k):
				break
			"""
			h: Graph = v.induce()
			j = h.colour_b(k - 1, i)
			if j == 0:
				#print("B: k = " + str(k) + " ret = 0")
				return 0
			i += j
			v.colour.value = i
			self.social_credit(v)
		max_degree: int = 0
		max_colour: int = 0
		for v in self.vertices():
			seen: Set[int] = set()
			e: Vertex
			for e in v.neighbours:
				seen.add(e.colour.value)
			if max_degree < v.degree():
				max_degree = v.degree()
			j = i
			while True:
				if j not in seen:
					v.colour.value = j
					if max_colour < j:
						max_colour = j
					j += 1
					break
				j += 1
		assert(max_colour < max_degree + i + 1)

		ret: int = max_colour - i + 1
		bound: float = 2 * k * Graph.magic_f(k, n)
		#print("Bend: k = {}, ret = {}, bound = {}".format(k, ret, bound))
		return 0 if ret > bound else ret

	def colour_c(self: "Graph") -> int:
		i: int = 1
		while self.duplicate().colour_b(1 << i, 1) == 0:
			i += 1
		#print("C: i = " + str(i))

		l: int = (1 << (i - 1)) + 1
		r: int = 1 << i
		while l < r:
			m: int = (l + r) // 2
			if self.duplicate().colour_b(m, 1) == 0:
				l = m + 1
			else:
				r = m
		k: int = self.duplicate().colour_b(l, 1)
		assert(k != 0)
		return k

	@staticmethod
	def from_file(path: str) -> "Graph":
		g: Graph = Graph()
		with open(path, encoding = 'utf-8') as f:
			m: Dict[int, Vertex] = {}
			for line in f:
				if line.startswith('#'):
					continue
				parts: List[str] = line.split()
				x: int = int(parts[0])
				y: int = int(parts[1])
				vx: Vertex = m.get(x)
				if vx is None:
					vx = Vertex()
					m[x] = vx
					g.shift(vx)
				vy: Vertex = m.get(y)
				if vy is None:
					vy = Vertex()
					m[y] = vy
					g.shift(vy)
				vx.neighbours.add(vy)
				vy.neighbours.add(vx)
		return g

def main(args: List[str]) -> None:
	sys.setrecursionlimit(1000 * 1000)
	resource.setrlimit(resource.RLIMIT_STACK, (128 * 1024 * 1024, resource.RLIM_INFINITY))
	#Graph.initialize()
	g: Graph = Graph.from_file(args[0])
	k: int = g.colour_c()
	print("k = " + str(k))
	print("iterations = " + str(iterations))
	assert(g.verify_colouring())

if __name__ == '__main__':
	main(sys.argv[1:])
